pkg/gatewayserver/io/ws/ws_test.go

```diff
@@ -1092,7 +1119,46 @@ func TestTraffic(t *testing.T) {
 						}
 					}
 
-				case lbslns.UplinkDataFrame, lbslns.JoinRequest:
+				case lbslns.UplinkDataFrame:
+					now := time.Now()
+					v.UpInfo.XTime = upXTime
+					v.UpInfo.RxTime = float64(now.Unix())
+					req, err := json.Marshal(v)
+					if err != nil {
+						panic(err)
+					}
+					if err := wsConn.WriteMessage(websocket.TextMessage, req); err != nil {
+						t.Fatalf("Failed to write message: %v", err)
+					}
+					select {
+					case up := <-gsConn.Up():
+						a.So(time.Since(*ttnpb.StdTime(up.Message.ReceivedAt)), should.BeLessThan, timeout)
+						up.Message.ReceivedAt = nil
+						var payload ttnpb.Message
+						a.So(lorawan.UnmarshalMessage(up.Message.RawPayload, &payload), should.BeNil)
+						if !a.So(&payload, should.Resemble, up.Message.Payload) {
+							t.Fatalf("Invalid RawPayload: %v", up.Message.RawPayload)
+						}
+						up.Message.RawPayload = nil
+						up.Message.RxMetadata[0].UplinkToken = nil
+						expectedUp := tc.ExpectedNetworkUpstream.(ttnpb.UplinkMessage)
+
+						// Set the correct xtime and timestamps for the assertion.
+						expectedUp.RxMetadata[0].Timestamp = timestamp
+						expectedUp.RxMetadata[0].Time = ttnpb.ProtoTime(&now)
+						expectedUp.RxMetadata[0].Time.Nanos = 0
```

Comment:
The nano second value is not used by the GS but is added in the expected message. I don't know if there is a more beautiful way to do this.